### DISCUSSION TITLE
Disable auto-upload by default

### DISCUSSION THREAD

**Discussion by smichel17** [streetcomplete/StreetComplete#3470] [Ideas]
**Use case**

I was planning on discussing this eventually anyway, but it's also timely now: in https://github.com/streetcomplete/StreetComplete/issues/3466#issuecomment-954281066, @westnordost wrote,

> Now, one can argue that the precise/coarse location of the user can be determined by
> 
> - his current contribution history (especially if he has auto-sync on)

- When we used overpass, auto-upload was very important, because some quests (house number) would not appear until after a pre-requisite quest (building type) had been uploaded. However, with offline data storage, this is no longer a requirement. Actually, it's the other way: some functionality is only available when auto-upload is off (undo of certain actions, like leaving notes).
- Auto-upload has a large impact on battery usage (https://github.com/streetcomplete/StreetComplete/discussions/3251).
- And it's probably the biggest privacy concern in the app.

**Proposed Solution:** Change the default setting to upload on wifi only.

**Risks:** Some people may not realize they need to upload now.

Maybe there could be a prompt, similar to when you are not logged in yet, if you reach a certain number of changes waiting to be uploaded or if you have changes which are older than a certain age. It would tell (or require?) the user to upload, and let them know about the auto-upload setting (with a warning about battery usage).

**Comment by Cj-Malone** [comment]
I don't know if this is expected behaviour, or a bug but I can only upload when the screen is on, when the screen goes off the upload stops until I unlock the phone (and have SC open?). If this is expected behaviour I'd argue for auto upload to remain the default, it's not a great experience.

**Reply by kmpoppe** [reply]
I was under the impression that this was fixed in one of the more recent releases. Are you working with v37?

**Reply by Cj-Malone** [reply]
Looking at the changesets that I last noticed it on it was 36.0. I'll check it out next time, hopefully it's sorted.

**Reply by Cj-Malone** [reply]
I've just done a survey with StreetComplete 37.0, and it did upload with the screen turned off. Thanks to everyone who made that happen.

As a side note, I saw quests split into multiple changesets which wasn't what I expected. For example "Add housenumbers" was [113733349](https://www.openstreetmap.org/changeset/113733349), [113736756](https://www.openstreetmap.org/changeset/113736756), [113739356](https://www.openstreetmap.org/changeset/113739356), [113740188](https://www.openstreetmap.org/changeset/113740188), [113741067](https://www.openstreetmap.org/changeset/113741067), [113741784](https://www.openstreetmap.org/changeset/113741784) and [113742966](https://www.openstreetmap.org/changeset/113742966).

I pressed upload around 1450, and didn't turn the screen back on until around 1950. It's not really an issue for me, but I though it was worth documenting.

**Comment by matkoniecz** [comment]
> Risks: Some people may not realize they need to upload now.

Also, it would be more annoying to edit both with SC and other editor like Vespucci at the same time.

**Reply by smichel17** [reply]
I'm not proposing removing the option, just changing the default.

**Reply by matkoniecz** [reply]
The problem is that vast majority of users never visits settings page.

**Reply by riQQ** [reply]
But people who would be using Vespucci and SC are advanced users and pretty likely will check out the options imo.

**Reply by matkoniecz** [reply]
1) SC is directed also to newbies

2) people here are extreme (and welcome and positive) outliers, there are thousands of users who never commented on GitHub

3) most of users never change any settings

4) it could be interesting to review edits and see how many people ever solved quest not enabled by default (other settings have effects that are not exposed)

**Reply by riQQ** [reply]
As already aluded to by smichel17: There has to be a prompt or something similar to ask the user if they want to turn on auto upload or if they prefer to do it manually. It's probably the best to show it after solving the first or the first 10  (or another sensible number) quests.

**Reply by mnalis** [reply]
> it could be interesting to review edits and see how many people ever solved quest not enabled by default (other settings have effects that are not exposed)

Unless my counting is wrong; if we ignore  `AddRoofShape` (which has just recently become non-default) , about `12258` out of `38807` SC users seems to have solved at least one **non-default quest** - that would be about `32%` of them, which sounds way too much. 
When I also remove `AddMaxSpeed` (which might have been default before? I don't know...) I get much more reasonable `4509`, which is more realistic `12%`; but it could be even lower if some of the quests below have been enabled by default before...  (let me know and I can recalculate).
 
```
% pbzip2 -dc changesets-211018.osm.bz2 | env -i LC_ALL=C fgrep -B1 'StreetComplete:quest_type'  | perl -nE '$u=$1 if /uid="(\d+)"/; say "$1\t$u" if /StreetComplete:quest_type" v="(.*?)"/' >  changesets_sc.txt

% sort -u changesets_sc.txt| cut -f2 | sort -u | wc -l
38807

% sort -u changesets_sc.txt | egrep -w 'AddBabyChangingTable|AddWheelchairAccessToiletsPart|AddAcceptsCash|AddVegetarian|AddVegan|AddHalal|AddKosher|AddWheelchairAccessBusiness|AddInternetAccess|AddMaxSpeed' | cut -f2 | sort -u | wc -l
12258
```

**Reply by smichel17** [reply]
`AddMaxSpeed` has been default-enabled in the past. I don't know about the others.

**Comment by mnalis** [comment]
For me, the biggest concern is whether new users would know they need to upload when they get to wifi. Even as an advanced user, I sometimes forget, as unless SC was the foreground app when I got home, `auto-upload on wifi` is equivalent to `off`. Sure, I'll remember in next day or two when I open SC again, but new user might not open it for some time (or never, if they don't see changes on osm.org they might decide SC is not working)...

Still, there are quite a few benefits, so I'm slightly more inclined to think that changing default is a good idea.

**Comment by smichel17** [comment]
> Even as an advanced user, I sometimes forget, as unless SC was the foreground app when I got home, `auto-upload on wifi` is equivalent to `off`

I agree this is the biggest concern; I also forget. I think a pre-requisite for mkaing this change would be showing a notification when you're back on wifi, like https://github.com/streetcomplete/StreetComplete/issues/1538#issuecomment-585356228

**Reply by mnalis** [reply]
Implementing that could be somewhat problematic, though, requiring that the app remains active in the background and do checks... 

Which would likely be problematic - at least with other background apps that really do **require** such periodic background checks (like DAVx5 for contacts/calendar sync for example), one needs to jump through many hoops, which is very unlikely most users will do (see https://dontkillmyapp.com/ for examples of workarounds one has to do to have any chance of periodic background checks working). And its getting worse and worse with every new android version :cry:

**Reply by smichel17** [reply]
As the primary maintainer of one such app through the time when Android was starting to implement these restrictions, I am very familiar with them. I have spent an unreasonable amount of time poring over the docs for [AlarmManager](https://developer.android.com/reference/android/app/AlarmManager) and related topics to figure out the best way to work around them, while still trying not to cause unnecessary battery drain (by the way, as a benevolent developer, this is by far the most frustrating part— that it is so complicated to do the best thing for users, while keeping your app working; it's so much easier to just `setAlarmClock` and be done with it.)

Anyway, StreetComplete is quite different, for three reasons:

1. The kind of app which is punished most harshly by Android's restrictions is one which runs near-**exclusively in the background**. At the time when I looked into it, if the user hadn't opened the app in the last 3 days, Android would stop delivering alarms at all (unless the user jumps through the hoops to exempt the app). The alert I proposed would only happen shortly (~24h or less) after using StreetComplete.

2. The second behavior which Android makes difficult is setting alerts which must happen at a **specific time** (and thus, may take the device out of a low-power state). Fortunately, StreetComplete doesn't care exactly when the alert happens. Actually, we'd *rather* have it happen when the user is actively using the device, so it's not missed. So, we can set generous parameters for when the alarm fires; the restrictions here don't really matter.

3. At least one of SC's alerts is in response to some other **trigger** (getting back on wifi), rather than being arbitrarily-scheduled. These have a different set of restrictions, which I am not as familiar with. I think (1) applies but not (2).

**Reply by riQQ** [reply]
You could also introduce a new user to this mechanism after they solve the first or the first 10 (or another sensible number) quests. Show a dialog and ask them if they want to turn auto upload on or if they want to do it manually by pushing the *upload changes* button.

**Comment by smichel17** [comment]
> There has to be a prompt or something similar to ask the user if they want to turn on auto upload or if they prefer to do it manually. It's probably the best to show it after solving the first or the first 10 (or another sensible number) quests.

@riQQ Note that such a mechanism already exists: the app prompts you to log in to OSM after some low number of changes (I think 3).

**Reply by riQQ** [reply]
So this is perfect to add the question / hint for auto upload / manual upload.

**Comment by FloEdelmann** [comment]
> And it's probably the biggest privacy concern in the app.

Disabling auto-upload (by default) might mitigate the problem described in #3507.

**Comment by timothywashere** [comment]
Slightly off topic but please bring back the option to turn off auto-download when using the app.
I have mobile data turned off and regularly get error messages as it keep trying to download quests. The error message appears on top of the quest dialog, so hard to ignore.

**Reply by matkoniecz** [reply]
That is quit weird and maybe a bug. With mobile data turned off it should not try to download anything.

Though it can be a bug in Android/phone OS.

Have you disable mobile data with phone-wide setting?

**Reply by timothywashere** [reply]
Yeah, I use the setting from the pull down menu although it's harder to get to since Android 12. Just turn the 4G toggle off and on when I need it.
Does SC cache the setting? IE if mobile data is on when I launch the app it keeps on checking?

**Reply by matkoniecz** [reply]
For me when I disable mobile data it shows error once (in case where some action was in progress) and is not pestering me with messages about failed attempts.

If mobile data is enabled but not working then I get repeated errors.

**Reply by Helium314** [reply]
Looks like Android is not telling the truth when checking connection using [connectivityManager.activeNetworkInfo](https://developer.android.com/reference/android/net/ConnectivityManager#getActiveNetworkInfo()).
